imap/mailbox.c

```diff
@@ -3178,15 +3189,24 @@ static void mailbox_quota_dirty(struct mailbox *mailbox)
     }
 }
 
+#define UPDATE_QUOTA_USED(quota_used, size, is_add)                     \
+    if (is_add) quota_used += size;                                     \
+    /* corruption prevention - check we don't go negative */            \
+    else if (quota_used > size) quota_used -= size;                     \
+    else quota_used = 0;
+
 static void header_update_counts(struct index_header *i,
                                  const struct index_record *record,
                                  int is_add)
 {
     int num = is_add ? 1 : -1;
 
     /* we don't track counts for EXPUNGED records */
-    if (record->internal_flags & FLAG_INTERNAL_EXPUNGED)
+    if (record->internal_flags & FLAG_INTERNAL_EXPUNGED) {
+        if (record->internal_flags & FLAG_INTERNAL_UNLINKED) is_add = !is_add;
```

Comment:
I'm not sure this logic is correct.  I think it should be at the top:

```
if (record->internal_flags & FLAG_INTERNAL_UNLINKED)
    return;
```

Because - this gets called twice, one with `is_add==0` for the old version of the record, and again with `is_add=1` for the new version.  So we will decrement the counter on the rewrite from EXPUNGED to EXPUNGED+UNLINKED like we should without this flag flip.

Reply:
Yeah, you're right.  Fixed